File: uhrwerk-backend-mysql/src/test/java/io/qimia/uhrwerk/dao/TargetDAOTest.java
package io.qimia.uhrwerk.dao;

import io.qimia.uhrwerk.common.metastore.config.TargetResult;
import io.qimia.uhrwerk.common.model.*;
import org.junit.jupiter.api.Test;

import java.sql.DriverManager;
import java.sql.SQLException;
import java.sql.Statement;

import static org.junit.jupiter.api.Assertions.*;

public class TargetDAOTest {

    java.sql.Connection db;

    @org.junit.jupiter.api.BeforeEach
    void setUp() throws SQLException {
        db = DriverManager.getConnection(
                "jdbc:mysql://localhost:53306/UHRWERK_METASTORE",
                "UHRWERK_USER",
                "Xq92vFqEKF7TB8H9"
        );
    }

    @org.junit.jupiter.api.AfterEach
    void tearDown() throws SQLException {
        // WARNING deletes all data as cleanup
        var deleteDependencyStm = db.createStatement();
        deleteDependencyStm.execute("DELETE FROM DEPENDENCY");
        deleteDependencyStm.close();
        var deleteSourceStm = db.createStatement();  // In case of some lost source data
        deleteSourceStm.execute("DELETE FROM SOURCE");
        deleteSourceStm.close();
        var deleteTargetStm = db.createStatement();
        deleteTargetStm.execute("DELETE FROM TARGET");
        deleteTargetStm.close();
        var deleteConnectionStm = db.createStatement();
        deleteConnectionStm.execute("DELETE FROM CONNECTION");
        deleteConnectionStm.close();
        var deleteTableStm = db.createStatement();
        deleteTableStm.execute("DELETE FROM TABLE_");
        deleteTableStm.close();
        if (db != null) if (!db.isClosed()) db.close();
    }

    Connection setupFSConnection() throws SQLException {
        Connection conn = new Connection();
        conn.setName("fsconn");
        conn.setPath("/aa/bb/cc/");
        conn.setType(ConnectionType.FS);
        conn.setKey();
        Statement setupTableStm = db.createStatement();
        setupTableStm.executeUpdate("INSERT INTO CONNECTION(id, name, type, path)" +
                "VALUES (" + conn.getId() + ", 'fsconn', 'FS', '/aa/bb/cc/')");
        setupTableStm.close();
        return conn;
    }

    Connection setupJDBCConnection() throws SQLException {
        Connection conn = new Connection();
        conn.setName("jdbcconn");
        conn.setJdbcPass("pass");
        conn.setJdbcUser("root");
        conn.setJdbcUrl("somejdbcurl");
        conn.setJdbcDriver("somejdbcdriver");
        conn.setType(ConnectionType.JDBC);
        conn.setKey();
        Statement setupTableStm = db.createStatement();
        setupTableStm.executeUpdate("INSERT INTO CONNECTION(id, name, type, jdbc_pass, jdbc_user, jdbc_url, " +
                "jdbc_driver) VALUES (" + conn.getId() + ", 'jdbcconn', 'JDBC', 'pass', 'root', 'somejdbcurl', " +
                "'somejdbcdriver')");
        setupTableStm.close();
        return conn;
    }

    Table setupTable() throws SQLException {
        var newTable = new Table();
        newTable.setArea("area1");
        newTable.setVertical("vertical1");
        newTable.setName("tablename");
        newTable.setVersion("1.0");
        newTable.setPartitionUnit(PartitionUnit.HOURS);
        newTable.setPartitionSize(1);
        newTable.setKey();
        Statement setupTableStm = db.createStatement();
        setupTableStm.executeUpdate("INSERT INTO TABLE_(id, area, vertical, name, version, partition_unit, partition_size, parallelism, max_bulk_size)" +
                "VALUES (" + newTable.getId() + ", 'area1', 'vertical1', 'tablename', '1.0', 'HOURS', 1, 1, 1)");
        setupTableStm.close();
        return newTable;
    }

    @Test
    void writeReadDelete() throws SQLException {
        Table t = setupTable();
        Connection fsConn = setupFSConnection();
        Target tarA = new Target();
        tarA.setFormat("parquet");
        tarA.setConnection(fsConn);
        tarA.setTableId(t.getId());
        tarA.setKey();

        Connection dbConn = setupJDBCConnection();
        Target tarB = new Target();
        tarB.setFormat("jdbc");
        tarB.setConnection(dbConn);
        tarB.setTableId(t.getId());
        tarB.setKey();

        TargetDAO dao = new TargetDAO(db);
        // write
        dao.insertTargets(new Target[]{tarA, tarB});

        // read
        Target[] retrievedTargets = dao.getTableTargets(t.getId());
        assertEquals(2, retrievedTargets.length);
        for (Target targetFound : retrievedTargets) {
            assertTrue((targetFound.getFormat().equals("parquet")) || (targetFound.getFormat().equals("jdbc")));
        }

        // delete
        dao.deleteTargetsByTable(t.getId());

        // read again
        retrievedTargets = dao.getTableTargets(t.getId());
        assertEquals(0, retrievedTargets.length);
    }

    @Test
    void fullSaveTest() throws SQLException {
        Table t = setupTable();
        Connection fsConn = setupFSConnection();
        Target tarA = new Target();
        tarA.setFormat("parquet");
        tarA.setConnection(fsConn);
        tarA.setTableId(t.getId());
        tarA.setKey();

        Connection dbConn = setupJDBCConnection();
        Target tarB = new Target();
        tarB.setFormat("jdbc");
        tarB.setConnection(dbConn);
        tarB.setTableId(t.getId());
        tarB.setKey();
        Target[] targets = {tarA, tarB};

        TargetDAO dao = new TargetDAO(db);
        TargetResult res1 = dao.save(targets, t.getId(), true);
        assertTrue(res1.isSuccess());

        dao.deleteTargetsByTable(t.getId());
        TargetResult res2 = dao.save(targets, t.getId(), false);
        assertTrue(res2.isSuccess());
        TargetResult res3 = dao.save(targets, t.getId(), false);
        assertTrue(res3.isSuccess());

        tarB.setFormat("badformatchange");
        TargetResult res4 = dao.save(targets, t.getId(), false);
        assertFalse(res4.isSuccess());
    }


    @Test
    void compareTest() {
        var connA = new Connection();
        connA.setName("connA");
        connA.setKey();
        var tarA = new Target();
        tarA.setFormat("jdbc");
        tarA.setTableId(123L);
        tarA.setConnection(connA);
        tarA.setKey();

        var connAFull = new Connection();
        connAFull.setName("connA");
        connAFull.setJdbcDriver("somedriver");
        connAFull.setJdbcUrl("someurl");
        connAFull.setJdbcUser("root");
        connAFull.setJdbcPass("somePass");
        var tarB = new Target();
        tarB.setFormat("jdbc");
        tarB.setTableId(123L);
        tarB.setConnection(connAFull);
        assertTrue(TargetDAO.compareTargets(tarB, tarA));

        tarA.setFormat("parquet");
        tarA.setKey();
        assertFalse(TargetDAO.compareTargets(tarB, tarA));

        tarA.setFormat("jdbc");
        connA.setName("newname");
        connA.setKey();
        tarA.setKey();
        assertFalse(TargetDAO.compareTargets(tarB, tarA));
    }
}

File: uhrwerk-backend-mysql/src/test/kotlin/io/qimia/uhrwerk/dao/TargetDAOTest.kt
package io.qimia.uhrwerk.dao

import com.google.common.truth.Truth
import io.qimia.uhrwerk.TestData
import TestUtils
import com.google.common.truth.Truth.assertThat
import io.qimia.uhrwerk.common.metastore.builders.ConnectionModelBuilder
import io.qimia.uhrwerk.common.metastore.builders.TargetModelBuilder
import io.qimia.uhrwerk.common.metastore.model.ConnectionModel
import io.qimia.uhrwerk.common.metastore.model.TableModel
import io.qimia.uhrwerk.common.model.*
import io.qimia.uhrwerk.dao.TargetDAO.Companion.compareTargets
import io.qimia.uhrwerk.repo.ConnectionRepo
import io.qimia.uhrwerk.repo.HikariCPDataSource
import io.qimia.uhrwerk.repo.TableRepo
import org.junit.jupiter.api.*
import org.slf4j.LoggerFactory
import org.testcontainers.containers.MySQLContainer
import org.testcontainers.containers.output.Slf4jLogConsumer
import org.testcontainers.junit.jupiter.Container
import org.testcontainers.junit.jupiter.Testcontainers
import java.sql.SQLException

@Testcontainers
class TargetDAOTest {

    private val service = TargetDAO()

    private var table: TableModel? = null
    private var connection: ConnectionModel? = null

    @BeforeEach
    @Throws(SQLException::class)
    fun addDeps() {
        connection = ConnectionRepo().save(TestData.connection("Connection-TargetTest"))
        table = TableRepo().save(TestData.table("Table-TargetTest"))
    }

    @AfterEach
    @Throws(SQLException::class)
    fun cleanUp() {
        TestUtils.cleanData("TARGET", LOGGER)
        TestUtils.cleanData("TABLE_", LOGGER)
        TestUtils.cleanData("CONNECTION", LOGGER)
    }

    @Test
    fun compareTest() {
        val connA = ConnectionModel()
        connA.name = "connA"

        val tarA = TargetModelBuilder()
            .format("jdbc")
            .tableId(123L)
            .connection(connA)
            .build()


        val connAFull = ConnectionModelBuilder()
            .name("connA")
            .jdbcDriver("somedriver")
            .jdbcUrl("someurl")
            .jdbcUser("root")
            .jdbcPass("somePass")
            .build()

        val tarB = TargetModelBuilder()
            .format("jdbc")
            .tableId(123L)
            .connection(connAFull)
            .build()

        Assertions.assertTrue(compareTargets(tarB, tarA))
        val tarA1 = TargetModelBuilder()
            .format("parquet")
            .tableId(123L)
            .connection(connA)
            .build()

        Assertions.assertFalse(compareTargets(tarB, tarA1))

        val connA2 = ConnectionModelBuilder()
            .name("newname").build()

        val tarA2 = TargetModelBuilder()
            .format("jdbc")
            .tableId(123L)
            .connection(connA2)
            .build()
        Assertions.assertFalse(compareTargets(tarB, tarA2))
    }

    @Test
    fun save() {
        val formats = listOf("lake", "csv")
        val targets = formats.map {
            TestData.target(
                table!!.id!!,
                connName = "Connection-TargetTest",
                format = it
            )
        }
        val result = service.save(targets, table!!.id!!, false)
        assertThat(result).isNotNull()
        assertThat(result.isSuccess).isTrue()
        assertThat(result.storedTargets).isNotNull()
        assertThat(result.storedTargets).isNotEmpty()
        assertThat(result.storedTargets!!.mapNotNull { it.format }.sorted())
            .isEqualTo(formats.sorted())
    }

    @Test
    fun saveSameNotOverwrite() {
        val formats = listOf("lake", "csv")
        val targets = formats.map {
            TestData.target(
                table!!.id!!,
                connName = "Connection-TargetTest",
                format = it
            )
        }
        val result = service.save(targets, table!!.id!!, false)
        assertThat(result).isNotNull()
        assertThat(result!!.isSuccess).isTrue()

        val sameTargets = formats.map {
            TestData.target(
                table!!.id!!,
                connName = "Connection-TargetTest",
                format = it
            )
        }
        val result1 = service.save(sameTargets, table!!.id!!, false)
        assertThat(result1).isNotNull()
        assertThat(result1.isSuccess).isTrue()

        assertThat(result1.storedTargets).isNotNull()
        assertThat(result1.storedTargets).isNotEmpty()

        assertThat(result1.storedTargets!!.mapNotNull { it.format }.sorted())
            .isEqualTo(formats.sorted())

        for (target in result1.storedTargets!!) {
            assertThat(result!!.storedTargets!!.toList()).contains(target)
        }
    }

    @Test
    fun newTargetNotOverwrite() {
        val formats = listOf("lake", "csv")
        val targets = formats.map {
            TestData.target(
                table!!.id!!,
                connName = "Connection-TargetTest",
                format = it
            )
        }
        val result = service.save(targets, table!!.id!!, false)

        val formats1 = listOf("lake", "csv", "parquet")
        val targets1 = formats1.map {
            TestData.target(
                table!!.id!!,
                connName = "Connection-TargetTest",
                format = it
            )
        }
        val result1 = service.save(targets1, table!!.id!!, false)
        assertThat(result1).isNotNull()
        assertThat(result1!!.isSuccess).isFalse()

        val targets2 = service.getTableTargets(table!!.id!!)
        assertThat(targets2).isNotNull()
        assertThat(targets2).isNotEmpty()

        for (target in targets2) {
            assertThat(result!!.storedTargets!!.toList()).contains(target)
        }
    }

    @Test
    fun newTargetOverwrite() {
        val formats = listOf("lake", "csv")
        val targets = formats.map {
            TestData.target(
                table!!.id!!,
                connName = "Connection-TargetTest",
                format = it
            )
        }
        service.save(targets, table!!.id!!, false)

        val formats1 = listOf("lake", "csv", "parquet")
        val targets1 = formats1.map {
            TestData.target(
                table!!.id!!,
                connName = "Connection-TargetTest",
                format = it
            )
        }
        val result1 = service.save(targets1, table!!.id!!, true)
        assertThat(result1).isNotNull()
        assertThat(result1!!.isSuccess).isTrue()

        val targets2 = service.getTableTargets(table!!.id!!)
        assertThat(targets2).isNotEmpty()
        assertThat(targets2).hasSize(formats1.size)
        for (target in targets2) {
            assertThat(result1!!.storedTargets!!.toList()).contains(target)
        }
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(TargetDAOTest::class.java)

        @Container
        var MY_SQL_DB: MySQLContainer<*> = TestUtils.mysqlContainer()

        @BeforeAll
        @JvmStatic
        fun setUp() {
            val logConsumer = Slf4jLogConsumer(LOGGER)
            MY_SQL_DB.followOutput(logConsumer)
            HikariCPDataSource.initConfig(
                MY_SQL_DB.jdbcUrl,
                MY_SQL_DB.username,
                MY_SQL_DB.password
            )
        }

        @AfterAll
        @JvmStatic
        fun tearDown() {
            HikariCPDataSource.close()
        }
    }

}